Error in start after delta upgrade of path which is having empty spaces

After doing an k2s system upgrade from an delta package in v1.8.1 we get an error in activating the cluster:

⏳⏳ [17:23:48] [Update] Applying artifacts from delta root 'C:\Program Files\k2s\1.8.1-from-1.8.0' to target 'C:\Program Files\k2s\1.8.0'                   ⏳⏳ [17:23:48] [Update][Relocate] New installation folder will be: C:\Program Files\k2s\1.8.1-from-1.8.0                                                    ⏳⏳ [17:23:48] [Update][Relocate] Previous installation retained for rollback: C:\Program Files\k2s\1.8.0                                                   ⏳⏳ [17:23:50] [Update] Seeding unchanged files from 'C:\Program Files\k2s\1.8.0' into 'C:\Program Files\k2s\1.8.1-from-1.8.0'                              ⏳⏳ [17:24:53] [Update] Seeding complete (robocopy exit code 3)                                                                                             ⏳ [17:24:53] [Update][Relocate] Removed obsolete file: lib/manifests/clusterip-webhook/certgen-patch-job.yaml
⏳ [17:24:53] [Update][Relocate] Removed obsolete file: lib/scripts/k2s/clusterip-webhook/Deploy-ClusterIPWebhook.ps1
⏳ [17:24:53] [Update][Relocate] Removed obsolete file: lib/manifests/clusterip-webhook/certgen-create-job.yaml
⏳ [17:24:53] [Update] Re-homing installation from 'C:\Program Files\k2s\1.8.0' to 'C:\Program Files\k2s\1.8.1-from-1.8.0'
⏳ [17:24:54] [Update] Re-pointed service 'containerd' parameter(s): AppDirectory, Application, AppParameters
⏳ [17:24:54] [Update] Re-pointed service 'flanneld' parameter(s): Application, AppParameters
⏳ [17:24:54] [Update] Re-pointed service 'kubelet' parameter(s): AppParameters
⏳ [17:24:54] [Update] Re-pointed service 'kubeproxy' parameter(s): AppDirectory, Application, AppParameters
⏳ [17:24:54] [Update] Re-pointed service 'dnsproxy' parameter(s): AppDirectory, Application, AppParameters
⏳ [17:24:54] [Update] Re-pointed service 'httpproxy' parameter(s): AppDirectory, Application
⏳ [17:24:54] [Update] Re-pointed 14 nssm service parameter(s) from 'C:\Program Files\k2s\1.8.0' to 'C:\Program Files\k2s\1.8.1-from-1.8.0'
⏳ [17:24:55] [Update] Updated StartKubelet.ps1 with new installation path
⏳ [17:24:55] Regenerated containerd config.toml for new installation path
⏳ [17:24:55] Updated machine PATH entries from 'C:\Program Files\k2s\1.8.0' to 'C:\Program Files\k2s\1.8.1-from-1.8.0'
⏳ [17:24:55] Copied kubeconfig from 'C:\Program Files\k2s\1.8.0\config' to 'C:\Program Files\k2s\1.8.1-from-1.8.0\config'
⏳ [17:24:55] setup.json InstallFolder set to 'C:\Program Files\k2s\1.8.1-from-1.8.0'
⏳ [17:24:55] [Relocate] Active installation folder is now: C:\Program Files\k2s\1.8.1-from-1.8.0
⏳ [17:24:55] Phase 8/16: CleanKubeletFlags
⏳ [17:24:55] Cleaning deprecated kubelet flags from C:\var\lib\kubelet\kubeadm-flags.env
⏳ [17:24:55] Removed deprecated --pod-infra-container-image flag
⏳ [17:24:55] Kubelet flags after cleanup: KUBELET_KUBEADM_ARGS="--hostname-override=WIN11TEST --node-ip=172.20.1.2 --config-dir=C:\\etc\\kubernetes\\kubelet.conf.d"
⏳ [17:24:55] Phase 9/16: RestartCluster
⏳ [17:24:55] Restarting K2s cluster...
⏳ [17:24:55] Starting cluster from the new installation folder...
⏳ [17:24:55] 🤖 Starting K2s on Microsoft Windows 11 Iot Enterprise Ltsc 2024 10.0.26100.1742 Build 26100.1742
⏳ [17:25:04] ⏳ [17:25:04] Configuring kubemaster VM
⏳ [17:26:34] ⏳ Write-ErrorMessage : [17:26:34][ERROR] Start-Service : Failed to start service 'dnsproxy (dnsproxy)'.  At C:\Program
⏳ [17:26:34] ⏳ Files\k2s\1.8.1-from-1.8.0\lib\modules\k2s\k2s.node.module\windowsnode\downloader\artifacts\nssm\nssm.module.psm1:127 char:13  +
⏳ [17:26:34] ⏳ Start-Service $Name -WarningAction SilentlyContinue  +             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~      + CategoryInfo          :
⏳ [17:26:34] ⏳ OpenError: (System.ServiceProcess.ServiceController:ServiceController) [Start-Service], ServiceCommandException      + FullyQualifiedErrorId :
⏳ [17:26:34] ⏳ StartServiceFailed,Microsoft.PowerShell.Commands.StartServiceCommand
⏳ [17:26:34] ⏳ At C:\Program Files\k2s\1.8.1-from-1.8.0\lib\modules\k2s\k2s.infra.module\log\log.module.psm1:293 char:21
⏳ [17:26:34] ⏳ + ...             Write-ErrorMessage -Message $message -ConsoleMessage $con ...
⏳ [17:26:34] ⏳ +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
⏳ [17:26:34] ⏳     + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
⏳ [17:26:34] ⏳     + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Write-ErrorMessage
⏳ [17:26:34] ⏳
⏳ [17:26:39] ⏳ [17:26:39] Starting windows services
⏳ [17:26:46] ⏳ [17:26:46] Starting Kubernetes services on the Windows node
⏳ [17:26:50] ⏳ Write-ErrorMessage : [17:26:50][ERROR] Start-Service : Failed to start service 'containerd (containerd)'.  At C:\Program
⏳ [17:26:50] ⏳ Files\k2s\1.8.1-from-1.8.0\lib\modules\k2s\k2s.node.module\windowsnode\downloader\artifacts\nssm\nssm.module.psm1:127 char:13  +
⏳ [17:26:50] ⏳ Start-Service $Name -WarningAction SilentlyContinue  +             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~      + CategoryInfo          :
⏳ [17:26:50] ⏳ OpenError: (System.ServiceProcess.ServiceController:ServiceController) [Start-Service], ServiceCommandException      + FullyQualifiedErrorId :
⏳ [17:26:50] ⏳ StartServiceFailed,Microsoft.PowerShell.Commands.StartServiceCommand
⏳ [17:26:50] ⏳ At C:\Program Files\k2s\1.8.1-from-1.8.0\lib\modules\k2s\k2s.infra.module\log\log.module.psm1:293 char:21
⏳ [17:26:50] ⏳ + ...             Write-ErrorMessage -Message $message -ConsoleMessage $con ...
⏳ [17:26:50] ⏳ +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
⏳ [17:26:50] ⏳     + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
⏳ [17:26:50] ⏳     + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Write-ErrorMessage
⏳ [17:26:50] ⏳
⏳ [17:26:58] ⏳ Write-ErrorMessage : [17:26:58][ERROR] Start-Service : Failed to start service 'kubeproxy (kubeproxy)'.  At C:\Program
⏳ [17:26:58] ⏳ Files\k2s\1.8.1-from-1.8.0\lib\modules\k2s\k2s.node.module\windowsnode\downloader\artifacts\nssm\nssm.module.psm1:127 char:13  +
⏳ [17:26:58] ⏳ Start-Service $Name -WarningAction SilentlyContinue  +             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~      + CategoryInfo          :
⏳ [17:26:58] ⏳ OpenError: (System.ServiceProcess.ServiceController:ServiceController) [Start-Service], ServiceCommandException      + FullyQualifiedErrorId :
⏳ [17:26:58] ⏳ StartServiceFailed,Microsoft.PowerShell.Commands.StartServiceCommand
⏳ [17:26:58] ⏳ At C:\Program Files\k2s\1.8.1-from-1.8.0\lib\modules\k2s\k2s.infra.module\log\log.module.psm1:293 char:21
⏳ [17:26:58] ⏳ + ...             Write-ErrorMessage -Message $message -ConsoleMessage $con ...
⏳ [17:26:58] ⏳ +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

It looks like in nssm a path like C:\Program Files\k2s\1.8.1-from-1.8.0\bin\dnsproxy.exe for the appplication does not work. Seems that the handling for delta upgrades is wring there when ⏳ [17:24:54] [Update] Re-pointed service 'dnsproxy' parameter(s): AppDirectory, Application, AppParameters

--- a/lib/modules/k2s/k2s.cluster.module/update/update.module.psm1
+++ b/lib/modules/k2s/k2s.cluster.module/update/update.module.psm1
@@ -592,9 +592,12 @@ function Set-K2sInstallationHome {
 					if ($current -is [Array]) { $current = ($current -join ' ') }
 					if ($current.IndexOf($fromTrim, [System.StringComparison]::OrdinalIgnoreCase) -lt 0) { continue }
 					$newValue = [regex]::Replace($current, $fromPattern, $toTrim, [System.Text.RegularExpressions.RegexOptions]::IgnoreCase)
-					& $nssmPath set $svc $parameter $newValue | Out-Null
-					if ($LASTEXITCODE -ne 0) {
-						Write-Log ("[Update][Warn] nssm fallback failed (exit {0}) re-pointing service '{1}' parameter '{2}'" -f $LASTEXITCODE, $svc, $parameter) -Console
+					# Write directly to the registry (see Update-NssmServiceInstallPath) instead of 'nssm.exe set'
+					# to avoid the native-command quoting bug for installation paths that contain spaces.
+					try {
+						Set-ItemProperty -Path $regKey -Name $parameter -Value $newValue -ErrorAction Stop
+					} catch {
+						Write-Log ("[Update][Warn] nssm fallback failed re-pointing service '{0}' parameter '{1}': {2}" -f $svc, $parameter, $_.Exception.Message) -Console
 						continue
 					}
 					$repointedCount++

--- a/lib/modules/k2s/k2s.node.module/windowsnode/services/services.module.psm1
+++ b/lib/modules/k2s/k2s.node.module/windowsnode/services/services.module.psm1
@@ -207,9 +207,17 @@ function Update-NssmServiceInstallPath {
 
         $newValue = [regex]::Replace($current, $oldPathPattern, $NewPath, [System.Text.RegularExpressions.RegexOptions]::IgnoreCase)
         Write-Log ("Re-pointing service '{0}' parameter '{1}' to new installation path" -f $Name, $parameter)
-        & $NssmPath set $Name $parameter $newValue | Out-Null
-        if ($LASTEXITCODE -ne 0) {
-            Write-Log ("[Update][Warn] nssm failed (exit {0}) re-pointing service '{1}' parameter '{2}'; not recording for rollback" -f $LASTEXITCODE, $Name, $parameter)
+        # Write the value DIRECTLY to the registry instead of via 'nssm.exe set'. nssm reads its
+        # configuration from these registry values, so a direct write is equivalent - and it avoids a
+        # PowerShell native-command quoting bug: when the installation path contains spaces (e.g.
+        # 'C:\Program Files\k2s'), an AppParameters value such as --config-path="C:\Program Files\...\x.yaml"
+        # mixes spaces with embedded double quotes. Passing that string as a native argument to nssm.exe
+        # makes PowerShell wrap it in quotes without escaping the embedded ones, so nssm stores a mangled
+        # value and the service fails to start. Set-ItemProperty writes the exact string verbatim.
+        try {
+            Set-ItemProperty -Path $regKey -Name $parameter -Value $newValue -ErrorAction Stop
+        } catch {
+            Write-Log ("[Update][Warn] Failed to write nssm registry value for service '{0}' parameter '{1}'; not recording for rollback: {2}" -f $Name, $parameter, $_.Exception.Message)
             continue
         }
         $previousValues[$parameter] = $current

--- a/lib/modules/k2s/k2s.node.module/windowsnode/services/services.module.unit.tests.ps1
+++ b/lib/modules/k2s/k2s.node.module/windowsnode/services/services.module.unit.tests.ps1
@@ -6,24 +6,6 @@ BeforeAll {
 	[Diagnostics.CodeAnalysis.SuppressMessageAttribute('UseDeclaredVarsMoreThanAssignments', '', Justification = 'Pester Test')]
 	$moduleName = (Import-Module "$PSScriptRoot\services.module.psm1" -PassThru -Force).Name
 	Mock -ModuleName $moduleName Write-Log {}
-
-	# Creates a platform-appropriate executable stub that exits with the given code, so the tests run
-	# on both Windows (.cmd) and Linux/macOS CI agents (.sh with a shebang + execute bit).
-	function New-NssmExitStub {
-		param(
-			[Parameter(Mandatory = $true)] [int] $ExitCode,
-			[Parameter(Mandatory = $true)] [string] $Directory
-		)
-		if ([System.Runtime.InteropServices.RuntimeInformation]::IsOSPlatform([System.Runtime.InteropServices.OSPlatform]::Windows)) {
-			$stub = Join-Path $Directory "nssm-stub-$ExitCode.cmd"
-			Set-Content -Path $stub -Value "@echo off`r`nexit /b $ExitCode" -NoNewline
-		} else {
-			$stub = Join-Path $Directory "nssm-stub-$ExitCode.sh"
-			Set-Content -Path $stub -Value "#!/bin/sh`nexit $ExitCode" -NoNewline
-			& chmod +x $stub
-		}
-		return $stub
-	}
 }
 
 Describe 'Update-NssmServiceInstallPath' -Tag 'unit', 'ci', 'update' {
@@ -46,20 +28,33 @@ Describe 'Update-NssmServiceInstallPath' -Tag 'unit', 'ci', 'update' {
 		Mock -ModuleName $moduleName Test-Path { $true } -ParameterFilter { $LiteralPath -like 'HKLM:\SYSTEM\CurrentControlSet\Services\*\Parameters' }
 		# AppDirectory is exactly the install root; without trimming, OldPath 'C:\k\' would not match it.
 		Mock -ModuleName $moduleName Get-ItemProperty { [pscustomobject]@{ AppDirectory = 'C:\k' } }
-		# Harmless real executable stub so the internal '& $NssmPath set ...' call exits cleanly.
-		$nssmStub = New-NssmExitStub -ExitCode 0 -Directory $TestDrive
-		$result = Update-NssmServiceInstallPath -Name 'containerd' -OldPath 'C:\k\' -NewPath 'C:\k2s\1.9.0\' -NssmPath $nssmStub
+		# The value is written directly to the registry via Set-ItemProperty; mock it so the test does not
+		# touch the real registry.
+		Mock -ModuleName $moduleName Set-ItemProperty { }
+		$result = Update-NssmServiceInstallPath -Name 'containerd' -OldPath 'C:\k\' -NewPath 'C:\k2s\1.9.0\' -NssmPath 'nssm.exe'
 		$result.Keys | Should -Contain 'AppDirectory'
 		$result['AppDirectory'] | Should -Be 'C:\k'
 	}
 
-	It 'does not record a parameter for rollback when nssm returns a non-zero exit code' {
+	It 'writes the re-pointed value verbatim to the registry (no native-command quoting) for paths with spaces' {
+		Mock -ModuleName $moduleName Get-Service { [pscustomobject]@{ Name = 'dnsproxy'; Status = 'Stopped' } }
+		Mock -ModuleName $moduleName Test-Path { $true } -ParameterFilter { $LiteralPath -like 'HKLM:\SYSTEM\CurrentControlSet\Services\*\Parameters' }
+		# AppParameters mixes spaces with embedded double quotes - the exact case that broke 'nssm.exe set'.
+		Mock -ModuleName $moduleName Get-ItemProperty { [pscustomobject]@{ AppParameters = '--config-path="C:\Program Files\k2s\1.8.0\bin\dnsproxy.yaml"' } }
+		$script:capturedValue = $null
+		Mock -ModuleName $moduleName Set-ItemProperty { $script:capturedValue = $Value } -ParameterFilter { $Name -eq 'AppParameters' }
+		$result = Update-NssmServiceInstallPath -Name 'dnsproxy' -OldPath 'C:\Program Files\k2s\1.8.0' -NewPath 'C:\Program Files\k2s\1.8.1-from-1.8.0' -NssmPath 'nssm.exe'
+		$result.Keys | Should -Contain 'AppParameters'
+		$script:capturedValue | Should -Be '--config-path="C:\Program Files\k2s\1.8.1-from-1.8.0\bin\dnsproxy.yaml"'
+	}
+
+	It 'does not record a parameter for rollback when the registry write fails' {
 		Mock -ModuleName $moduleName Get-Service { [pscustomobject]@{ Name = 'containerd'; Status = 'Stopped' } }
 		Mock -ModuleName $moduleName Test-Path { $true } -ParameterFilter { $LiteralPath -like 'HKLM:\SYSTEM\CurrentControlSet\Services\*\Parameters' }
 		Mock -ModuleName $moduleName Get-ItemProperty { [pscustomobject]@{ AppDirectory = 'C:\k' } }
-		# nssm stub that fails (e.g. ACL-protected registry key / locked service): exits non-zero.
-		$nssmStub = New-NssmExitStub -ExitCode 1 -Directory $TestDrive
-		$result = Update-NssmServiceInstallPath -Name 'containerd' -OldPath 'C:\k' -NewPath 'C:\k2s\1.9.0' -NssmPath $nssmStub
+		# Registry write fails (e.g. ACL-protected key / locked service).
+		Mock -ModuleName $moduleName Set-ItemProperty { throw 'access denied' }
+		$result = Update-NssmServiceInstallPath -Name 'containerd' -OldPath 'C:\k' -NewPath 'C:\k2s\1.9.0' -NssmPath 'nssm.exe'
 		# The failed parameter must NOT be reported as updated, so the caller does not believe
 		# the relocation succeeded for a service that still points at the old path.
 		$result.Keys | Should -Not -Contain 'AppDirectory'

